sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c

```diff
@@ -528,29 +534,75 @@ bool calc_fast9_lukas_kanade(struct opticflow_t *opticflow, struct image_t *img,
 
   // TODO scale flow to rad/s here
 
+  // ***************
   // Flow Derotation
+  // ***************
+
   float diff_flow_x = 0.f;
   float diff_flow_y = 0.f;
 
   if (opticflow->derotation && result->tracked_cnt > 5) {
-    diff_flow_x = (opticflow->img_gray.eulers.phi - opticflow->prev_img_gray.eulers.phi) * OPTICFLOW_FX;
-    diff_flow_y = (opticflow->img_gray.eulers.theta - opticflow->prev_img_gray.eulers.theta) * OPTICFLOW_FY;
-    /*diff_flow_x = (cam_state->rates.p)  / result->fps * img->w /
-                  OPTICFLOW_FOV_W;// * img->w / OPTICFLOW_FOV_W;
-    diff_flow_y = (cam_state->rates.q) / result->fps * img->h /
-                  OPTICFLOW_FOV_H;// * img->h / OPTICFLOW_FOV_H;*/
+
+    // determine the roll, pitch, yaw differencces between the images.
+    float phi_diff = opticflow->img_gray.eulers.phi - opticflow->prev_img_gray.eulers.phi;
+    float theta_diff = opticflow->img_gray.eulers.theta - opticflow->prev_img_gray.eulers.theta;
+    float psi_diff = opticflow->img_gray.eulers.psi - opticflow->prev_img_gray.eulers.psi;
+
+    if (strcmp(OPTICFLOW_CAMERA.dev_name, "/dev/video0") == 0) {
+      // bottom cam: just subtract a scaled version of the roll and pitch difference from the global flow vector:
+      diff_flow_x = phi_diff * OPTICFLOW_FX;
+      diff_flow_y = theta_diff * OPTICFLOW_FY;
+      //diff_flow_x = (cam_state->rates.p)  / result->fps * img->w / OPTICFLOW_FOV_W;// * img->w / OPTICFLOW_FOV_W;
+      //diff_flow_y = (cam_state->rates.q) / result->fps * img->h / OPTICFLOW_FOV_H;// * img->h / OPTICFLOW_FOV_H;
+    } else {
+      // for frontal cam, predict individual flow vectors:
+      struct flow_t *predicted_flow_vectors = predict_flow_vectors(vectors, result->tracked_cnt, phi_diff, theta_diff,
+                                              psi_diff, opticflow);
+      if (opticflow->show_flow) {
+        uint8_t color[4] = {255, 255, 255, 255};
+        uint8_t bad_color[4] = {255, 255, 255, 255};
+        image_show_flow_color(img, predicted_flow_vectors, result->tracked_cnt, opticflow->subpixel_factor, color, bad_color);
+      }
+
+      for (int i = 0; i < result->tracked_cnt; i++) {
+        // subtract the flow:
+        vectors[i].flow_x -= predicted_flow_vectors[i].flow_x;
+        vectors[i].flow_y -= predicted_flow_vectors[i].flow_y;
+      }
+    }
   }
 
   float rotation_threshold = M_PI / 180.0f;
+
```

Comment:
The code below and above look like they should be rearranged slightly. I would propose, first check that there are enough corners and that the derotation can be applied then apply it with an if else for the camera type. You will have to check and confirm if this will have the desired behaviour.

Reply:
Yes, will do.